libshpool/src/daemon/server.rs

```diff
@@ -732,27 +743,51 @@ impl Server {
         // spawn a background thread to reap the shell when it exits
         // and notify about the exit by closing a channel.
         let child_exit_notifier = Arc::new(ExitNotifier::new());
-        let waitable_child = fork.clone();
+
+        // Just cloning the fork and directly calling wait_for_exit() on it would
+        // be simpler, but it would be wrong because then the destructor for the
+        // cloned fork object would close the pty fd earlier than we want as the
+        // child watcher thread exits. This can cause the shell->client thread
+        // to read the wrong file (for example, the config file contents if the
+        // config watcher reloads).
+        let waitable_child_pid = fork.child_pid().ok_or(anyhow!("missing child pid"))?;
```

Comment:
Try to make sure I understand the case correctly: this is actually a bug in the shpool_pty crate that the Fork object can be cloned but it shouldn't be, as the resulting cloned object could close the underlying pty fd without notifying the original object, thus causing use-after-free errors. Is that right?

I think it'd be easier to understand if we highlight in the comment that the Fork object logically has 2 parts: the child pid and the automatically managed pty fd (via `Drop`). The child watcher thread only needs the unmanaged child pid part.

Reply:
That makes sense. I added a paragraph trying to make the distinction between the two parts a little clearer.

Reply:
Thanks, that helps!